sync_diff_inspector/diff.go

```diff
@@ -49,118 +51,277 @@ type Diff struct {
 }
 
 // NewDiff returns a Diff instance.
-func NewDiff(ctx context.Context, db1, db2 *sql.DB, cfg *Config) (diff *Diff, err error) {
+func NewDiff(ctx context.Context, cfg *Config) (diff *Diff, err error) {
 	diff = &Diff{
-		db1:              db1,
-		db2:              db2,
-		schema:           cfg.SourceDBCfg.Schema,
+		sourceDBs:        make(map[string]DBConfig),
 		chunkSize:        cfg.ChunkSize,
 		sample:           cfg.Sample,
 		checkThreadCount: cfg.CheckThreadCount,
 		useRowID:         cfg.UseRowID,
-		tables:           cfg.Tables,
+		useChecksum:      cfg.UseChecksum,
+		tables:           make(map[string]map[string]*TableConfig),
 		sqlCh:            make(chan string),
-		report:           NewReport(cfg.SourceDBCfg.Schema),
+		report:           NewReport(),
 		ctx:              ctx,
 	}
-	for _, table := range diff.tables {
-		table.Info, err = dbutil.GetTableInfoWithRowID(ctx, diff.db1, diff.schema, table.Name, cfg.UseRowID)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		table.Schema = diff.schema
-	}
-	diff.fixSQLFile, err = os.Create(cfg.FixSQLFile)
-	if err != nil {
+
+	if err = diff.init(cfg); err != nil {
+		diff.Close()
 		return nil, errors.Trace(err)
 	}
 
 	return diff, nil
 }
 
-// Equal tests whether two database have same data and schema.
-func (df *Diff) Equal() (err error) {
-	defer func() {
-		df.fixSQLFile.Close()
-		df.db1.Close()
-		df.db2.Close()
-	}()
+func (df *Diff) init(cfg *Config) (err error) {
+	// create connection for source.
+	if err = df.CreateDBConn(cfg); err != nil {
+		return errors.Trace(err)
+	}
 
-	df.wg.Add(1)
-	go func() {
-		df.WriteSqls()
-		df.wg.Done()
-	}()
+	if err = df.AdjustTableConfig(cfg); err != nil {
+		return errors.Trace(err)
+	}
 
-	tbls1, err := dbutil.GetTables(df.ctx, df.db1, df.schema)
+	df.fixSQLFile, err = os.Create(cfg.FixSQLFile)
 	if err != nil {
-		err = errors.Trace(err)
-		return
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+func (df *Diff) CreateDBConn(cfg *Config) (err error) {
+	for _, source := range cfg.SourceDBCfg {
+		source.Conn, err = dbutil.OpenDB(source.DBConfig)
+		if err != nil {
+			return errors.Errorf("create source db %+v error %v", source.DBConfig, err)
+		}
+		source.Conn.SetMaxOpenConns(cfg.CheckThreadCount)
```

Comment:
set this to avoid error like `dial tcp 10.26.2.1:3306: connect: cannot assign requested address`

Reply:
It’s better to leave comments in code files